platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/ConsensusImpl.java

```diff
@@ -16,1037 +16,751 @@
+        final ElectionRound electionRound = rounds.getElectionRound();
+        if (electionRound.isDecided() && noInitJudgesMissing()) {
+            // all famous witnesses for this round are now known. None will ever be added again. We
+            // know this round has at least one witness. We know they all have fame decided. We
+            // know the next 2 rounds have events in them, because otherwise we couldn't have
+            // decided the fame here. Therefore, any new witness added to this round in the future
+            // will be instantly decided as not famous. Therefore, the set of famous witnesses
+            // in this round is now completely known and immutable. So we can call the following, to
+            // record that fact, and propagate appropriately.
+            return roundDecided(electionRound);
+        }
+        return null;
+    }
 
-        // check if it's a witness. If so, vote in all elections in the current round, put new consensus events in nce
-        final List<EventImpl> newConsensusEvents = vote(event, roundInfo, stronglySeen);
-        logger.info(
-                ADD_EVENT.getMarker(),
-                "Event to be added: {} for round: {}, " + "stronglySeen: {}, newConsensusEvents: {}",
-                () -> EventUtils.toShortString(event),
-                () -> roundInfo.getRound(),
-                () -> EventUtils.toShortStrings(stronglySeen),
-                () -> EventUtils.toShortStrings(newConsensusEvents));
-
-        return newConsensusEvents;
+    /**
+     * @return true if there are no init judges missing
+     */
+    private boolean noInitJudgesMissing() {
+        return initJudges == null || initJudges.allJudgesFound();
     }
 
     /**
-     * Check if event is a witness, and if it is, make it vote in all elections in its created round. This should be
-     * called on a round R only if it has already been called on round R-1.
+     * Checks if an event is an init judge. If it is, it will set its round created and judge flags.
+     * if it's the last missing judge, it will also mark events which have previously reached
+     * consensus.
      *
-     * @param event
-     * 		the event that will vote (if it's a witness)
-     * @param roundInfo
-     * 		the RoundInfo for the created round of event.
-     * @param stronglySeen
-     * 		a list of all the witnesses in the previous round that it strongly sees
-     * @return list of all events that reach consensus during this method call, in consensus order (or null if none)
+     * @param event the event to check
      */
-    private List<EventImpl> vote(EventImpl event, RoundInfo roundInfo, ArrayList<EventImpl> stronglySeen) {
-        if (!event.isWitness()) { // only witnesses are allowed to vote
-            return null;
+    private void checkInitJudges(@NonNull final EventImpl event) {
+        if (noInitJudgesMissing() || !initJudges.isInitJudge(event.getBaseHash())) {
```

Comment:
If this execution condition was moved as a check before the method call, we'd save on the cost of making the function call.  Assuming the method is called in one or two placed.  But the optimization would be counting pennies.